src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Razor.Language.Syntax
+{
+    internal static class MarkupElementRewriter
+    {
+        public static RazorSyntaxTree Rewrite(RazorSyntaxTree syntaxTree)
+        {
+            var rewriter = new Rewriter();
+            var rewrittenRoot = rewriter.Visit(syntaxTree.Root);
+
+            var newSyntaxTree = RazorSyntaxTree.Create(rewrittenRoot, syntaxTree.Source, syntaxTree.Diagnostics, syntaxTree.Options);
+            return newSyntaxTree;
+        }
+
+        public static RazorSyntaxTree RemoveMarkupElement(RazorSyntaxTree syntaxTree)
+        {
+            var remover = new Remover();
+            var rewrittenRoot = remover.Visit(syntaxTree.Root);
+
+            var newSyntaxTree = RazorSyntaxTree.Create(rewrittenRoot, syntaxTree.Source, syntaxTree.Diagnostics, syntaxTree.Options);
+            return newSyntaxTree;
+        }
+
+        private class Rewriter : SyntaxRewriter
+        {
+            // From http://dev.w3.org/html5/spec/Overview.html#elements-0
+            private static readonly HashSet<string> VoidElements = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+            {
+                "area",
+                "base",
+                "br",
+                "col",
+                "command",
+                "embed",
+                "hr",
+                "img",
+                "input",
+                "keygen",
+                "link",
+                "meta",
+                "param",
+                "source",
+                "track",
+                "wbr"
+            };
+
+            private readonly Stack<TagBlockTracker> _startTagTracker = new Stack<TagBlockTracker>();
+
+            private TagBlockTracker CurrentTracker => _startTagTracker.Count > 0 ? _startTagTracker.Peek() : null;
+
+            private string CurrentStartTagName => CurrentTracker?.TagName;
+
+            public override SyntaxNode Visit(SyntaxNode node)
+            {
+                if (node != null)
+                {
+                    node = RewriteNode(node);
+                }
+
+                return base.Visit(node);
```

Comment:
Does it make sense for this to be a preorder traversal instead? https://github.com/aspnet/Blazor/blob/master/src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentDocumentRewritePass.cs#L55

Reply:
I'll take a look

Reply:
That works as well